### PR TITLE
CAT-2043 Translate toast messages

### PR DESCRIPTION
**Pull request by ndrnour**

### CODE REVIEW

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by ndrnour** [comment]
please run espresso tests

File: catroid/src/main/java/org/catrobat/catroid/utils/ToastUtil.java [hunk 1]
```diff
@@ -39,18 +39,10 @@ public static void showError(Context context, String message) {
...
-	}
-
 	public static void showSuccess(Context context, String message) {
 		createToast(context, message, false);
 	}
 
-	public static void showSuccess(Context context, int messageId) {
-		createToast(context, context.getResources().getString(messageId), false);
-	}
-
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/utils/ToastUtil.java]: Ahm, I'm not sure about these changes. Why did you remove these methods?

Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/utils/ToastUtil.java]: just unused Methods.

Comment (robertpainsi) [reply catroid/src/main/java/org/catrobat/catroid/utils/ToastUtil.java]: Yes :smile: My question _Why did you remove these methods?_ was meant to be much more general. So I also implied changes like
```diff
-ToastUtil.showError(…, R.string.foo);
+ToastUtil.showError(…, getString(R.string.foo))
```
in my statement. imo this helper method is definitely useful because you already pass in a context object (from which you get your resources).

___
Btw, please take a look at the hint about RTL at https://jira.catrob.at/browse/CAT-2043?focusedCommentId=18602&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-18602

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by thmq** [comment]
test this please

**Comment by thmq** [comment]
please run espresso tests

**Comment by thmq** [comment]
please run espresso tests

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/RecyclerViewFragment.java [hunk 2]
```diff
@@ -296,7 +296,7 @@ protected void prepareActionMode(@ActionModeType int type) {
 
 	private void startActionMode(@ActionModeType int type) {
 		if (adapter.getItems().isEmpty()) {
-			ToastUtil.showError(getActivity(), R.string.am_empty_list);
+			ToastUtil.showError(getActivity(), getString(R.string.am_empty_list));
```
[the remark this answers is not in the document]
Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/RecyclerViewFragment.java]: cause we have 2 methods which they do the same purpose as far as i know
`public static void showError(Context context, String message) {
		createToast(context, message, true);
	}`

`public static void showError(Context context, int messageId) {
		createToast(context, context.getResources().getString(messageId), true);
	}`
so i just used the first method everywhere, the second one is now unused.

[the remark this answers is not in the document]
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/RecyclerViewFragment.java]: ok 👍 shouldn't we then remove the unused method? I don't know, what do you think?

[the remark this answers is not in the document]
Comment (robertpainsi) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/RecyclerViewFragment.java]: IMO both methods are useful. They are util methods to provide an easy way to create a Toast, either by a string (which can be used for dynamic data) or a string id. The string id method has the same purpose but provides a shorter, simpler way to create a Toast which is a perfect use case for a util method.

It simplifies lots of cases but mostly cases like
```diff
- ToastUtil.showError(currentActivity, R.string.error_scratch_program_conversion_failed);
+ ToastUtil.showError(currentActivity, currentActivity.getResources().getString(R.string.error_scratch_program_conversion_failed));
```

btw, [Googles Toast.makeText](https://developer.android.com/reference/android/widget/Toast.html) provides two ways to create a standard Toast as well (string id or string).
___
For this ticket, please only address the hard-coded strings and string concatenation (localization) for now. We can talk about removing (or improving) the util methods later :wink:
```java
// $ grep -ER 'ToastUtil.*"'
ToastUtil.showError(PreStageActivity.this, "Error: connecting to " + host + ":" + port + " failed");
ToastUtil.showError(this, getResources().getText(R.string.error_project_upload).toString() + " " + serverAnswer);
ToastUtil.showError(this, R.string.notification_low_battery + " " + Integer.toString(value));
ToastUtil.showSuccess(context, numberOfBackPackedItems + " " + textForScripts + " "
ToastUtil.showSuccess(activity, selectedScriptGroupBackPack + " " + textForUnPacking);
ToastUtil.showError(getActivity(), "There was a problem during Google+ Signin. Status: "
```

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by ndrnour** [comment]
please run espresso tests

**Comment by ndrnour** [comment]
@robertpainsi  could you please review again

File: catroid/src/main/java/org/catrobat/catroid/drone/ardrone/DroneStageActivity.java [hunk 3]
```diff
@@ -147,7 +147,7 @@ public void onDroneBatteryChanged(int value) {
 
 		DroneControlService dcs = DroneServiceWrapper.getInstance().getDroneService();
 		if (dcs != null && (value < DroneInitializer.DRONE_BATTERY_THRESHOLD) && dcs.getDroneNavData().flying && !droneBatteryMessageShown) {
-			ToastUtil.showError(this, R.string.notification_low_battery + " " + Integer.toString(value));
+			ToastUtil.showError(this, getString(R.string.notification_low_battery) + " " + value);
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/drone/ardrone/DroneStageActivity.java]: If you want, you can also localize this resource. Up to you :wink:

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 4]
```diff
@@ -199,7 +199,7 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
 					intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP);
 					startActivity(intent);
 				} else {
-					ToastUtil.showError(getApplicationContext(), R.string.error_no_whatsapp);
+					ToastUtil.showError(getBaseContext(), R.string.error_no_whatsapp);
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: I can't approve these changes. For the ToastUtil, I guess it doesn't matter which context you pass in?…

@thmq, since you're very experienced with Android, maybe you can take a look at the context changes?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java [hunk 5]
```diff
@@ -342,7 +342,7 @@ private void triggerGPlusLogin(GoogleSignInResult result) {
 			GoogleSignInAccount account = result.getSignInAccount();
 			onGoogleLogInComplete(account);
 		} else {
-			ToastUtil.showError(getActivity(), "ERROR Google+ Signin. Status: " + result.getStatus());
+			ToastUtil.showError(getActivity(), getString(R.string.error_google_plus_sign_in, result.getStatus().getStatusMessage()));
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java]: Even though the documentation for the Status class is non existent, let's go with that.

File: catroid/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -901,6 +901,8 @@
     <string name="show_password">Show Password</string>
     <string name="register_password_mismatch">Passwords do not match</string>
     <string name="sign_in_error">Error during Sign-In</string>
+    <string name="error_connecting_to">Error: connecting failed to %s : %d%%</string>
```
Comment (robertpainsi) [inline catroid/src/main/res/values/strings.xml]: Either use
* `Failed to connect to %s : %d`
* `Connection to %s : %d failed`
* `Couldn't connect to %s : %d`
* or something similar

I could swear that the [material design writing guidelines](https://material.io/design/usability/accessibility.html#writing) have been much more precise on such strings. From a consistency point of view it doesn't matter which one you choose, we use all styles :smile:

IMO it should be `Couldn't connect to Raspberry Pi` (optionally with host and port). What do you think?

(Adding/changing string resources is expensive. It takes a long time until strings are translated to all languages. So we should always choose strings carefully!)

Comment (robertpainsi) [reply catroid/src/main/res/values/strings.xml]: btw, is there a better place for that string? All other strings in this block deal with login user accounts (login, register) only.

File: catroid/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -901,6 +901,8 @@
...
+    <string name="error_google_plus_sign_in">ERROR Google+ Sign in. Status: %s</string>
```
Comment (robertpainsi) [inline catroid/src/main/res/values/strings.xml]: Maybe you have some example status messages (since you used `result.getStatus().getStatusMessage()`)?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java (hunk 5 above)
Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java]: `result.getStatus()` will return an object of type Status
but `result.getStatus().getStatusMessage()` will return a string therefor i used it.
did i do it wrong?

File: catroid/src/main/java/org/catrobat/catroid/drone/ardrone/DroneStageActivity.java (hunk 3 above)
Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/drone/ardrone/DroneStageActivity.java]: ok

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java (hunk 4 above)
Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: i used `getBaseContext()` instead of `getApplicationContext()` to reduce using the _**static_Contetxt**_

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java (hunk 5 above)
Comment (robertpainsi) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/login/SignInDialog.java]: > did i do it wrong?

Quite the opposite actually! :wink: I like your solution.

I was just mentioning that there is no [documentation for the getStatusMessage method](https://developers.google.com/android/reference/com/google/android/gms/common/api/Status#getStatusMessage()). I also couldn't find more information online. There are some code snippets at https://developers.google.com/games/services/android/signin which are using `getStatusMessage` as well. However, it still would've been great to see some example messages that `getStatusMessage` return.

I can't say for sure what `result.getStatus().toString()` returns but it looks like it's returning a String only meant to be read by developers. `getStatusMessage` seems better suited. So, let's go with your solution! :1st_place_medal:

File: catroid/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1366,6 +1367,7 @@
     <!-- Raspi Settings -->
     <string name="raspi_settings_title_name">Raspberry Pi settings</string>
     <string name="raspi_settings_gpio_version">GPIO Version</string>
+    <string name="error_connecting_to">Error: Could not connect to %s : %d%%</string>
```
Comment (robertpainsi) [inline catroid/src/main/res/values/strings.xml]: Ahm, why the two `%` at the end?

Comment (ndrnour) [reply catroid/src/main/res/values/strings.xml]: i was getting a lint error without %% .
but now it's fine -_- i don't know how or why! just a magic happened :)))

**Comment by ndrnour** [comment]
please run espresso tests

File: catroid/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -901,6 +901,7 @@
     <string name="show_password">Show Password</string>
     <string name="register_password_mismatch">Passwords do not match</string>
     <string name="sign_in_error">Error during Sign-In</string>
+    <string name="error_google_plus_sign_in">ERROR Google+ Sign in. Status: %s</string>
```
Comment (robertpainsi) [inline catroid/src/main/res/values/strings.xml]: After looking through a lot of string resources, I think there is no value in trying to improve these newly added strings :laughing: So I'll keep my suggestions at a minimum. Feel free to add feedback, especially if you disagree.

`Google Sign-In error. Status: %d`

Comment (ndrnour) [reply catroid/src/main/res/values/strings.xml]: why %d at  end?

Comment (robertpainsi) [reply catroid/src/main/res/values/strings.xml]: > why %d at end? string not Integer

Ah, yes! Nice catch!
